### PR TITLE
Adding in ordered bufferable generation for the primitive types

### PR DESCRIPTION
**Pull request by ianoc**
Starting to add some more of these ordered bufferables

### CODE REVIEW

File: scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/PrimitiveOrderedBuf.scala [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import scala.language.experimental.macros
+import scala.reflect.macros.Context
+
+import com.twitter.scalding._
+import java.nio.ByteBuffer
+import com.twitter.scalding.typed.OrderedBufferable
+
+object GenPrimitive {
+  def dispatch(c: Context): PartialFunction[c.Type, TreeOrderedBuf[c.type]] = {
+    case tpe if tpe =:= c.universe.typeOf[Short] => PrimitiveOrderedBuf(c)(tpe, "getShort", "putShort")
```
Comment (johnynek) [inline scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/PrimitiveOrderedBuf.scala]: what about byte?

File: scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/PrimitiveOrderedBuf.scala [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    case tpe if tpe =:= c.universe.typeOf[Int] => PrimitiveOrderedBuf(c)(tpe, "getInt", "putInt")
+    case tpe if tpe =:= c.universe.typeOf[Long] => PrimitiveOrderedBuf(c)(tpe, "getLong", "putLong")
+    case tpe if tpe =:= c.universe.typeOf[Float] => PrimitiveOrderedBuf(c)(tpe, "getFloat", "putFloat")
+    case tpe if tpe =:= c.universe.typeOf[Double] => PrimitiveOrderedBuf(c)(tpe, "getDouble", "putDouble")
```
Comment (johnynek) [inline scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/PrimitiveOrderedBuf.scala]: do you want a `case _ => abort...`

File: scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/TreeOrderedBuf.scala [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  }
+}
+abstract class TreeOrderedBuf[C <: Context] {
+  val ctx: C
+  val tpe: ctx.Type
+  // Expected byte buffers to be in values a and b respestively, the tree has the value of the result
+  def compareBinary: ctx.Tree // ctx.Expr[Function2[ByteBuffer, ByteBuffer, Int]]
+  // expects the thing to be tested on in the indiciated TermName
+  def hash: (ctx.TermName, ctx.Tree)
+  def classz: ctx.Tree // ctx.Expr[Class[T]]
```
Comment (johnynek) [inline scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/TreeOrderedBuf.scala]: not needed.

File: scalding-macros/src/test/scala/com/twitter/scalding/macros/MacroOrderingProperties.scala [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    buf.position(0)
+    buf
+  }
+
+  def rawCompare[T](a: T, b: T)(implicit obuf: OrderedBufferable[T]): Int = {
+    obuf.compareBinary(serialize(a), serialize(b)).unsafeToInt
+  }
+
+  def oBufCompare[T](a: T, b: T)(implicit obuf: OrderedBufferable[T]): Int = {
+    obuf.compare(a, b)
```
Comment (johnynek) [inline scalding-macros/src/test/scala/com/twitter/scalding/macros/MacroOrderingProperties.scala]: do we need to clamp this?

File: scalding-macros/src/test/scala/com/twitter/scalding/macros/MacroOrderingProperties.scala [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    i match {
+      case x if x < 0 => -1
+      case x if x > 0 => 1
+      case x => 0
+    }
+  def check[T: Arbitrary](implicit ord: Ordering[T], obuf: OrderedBufferable[T]) = forAll { (a: T, b: T) =>
+    rawCompare(a, b) + rawCompare(b, a) shouldEqual 0
+    oBufCompare(a, b) + oBufCompare(b, a) shouldEqual 0
+    assert(rawCompare(a, b) === clamp(ord.compare(a, b)))
+    assert(oBufCompare(a, b) === ord.compare(a, b))
```
Comment (johnynek) [inline scalding-macros/src/test/scala/com/twitter/scalding/macros/MacroOrderingProperties.scala]: do we need clamp here?

File: scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/TreeOrderedBuf.scala [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+        private[this] def innerCompare(a: _root_.java.nio.ByteBuffer, b: _root_.java.nio.ByteBuffer): Int = {
+            val ${t.compareBinary._1} = a
+            val ${t.compareBinary._2} = b
+            ${t.compareBinary._3}
+            return 0
+          }
+
+        def compareBinary(a: _root_.java.nio.ByteBuffer, b: _root_.java.nio.ByteBuffer): _root_.com.twitter.scalding.typed.OrderedBufferable.Result = {
+            val r = innerCompare(a, b)
```
Comment (johnynek) [inline scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/TreeOrderedBuf.scala]: need the try { } catch here, I think.

File: scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/TreeOrderedBuf.scala [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+          }
+
+        def hash(passedInObjectToHash: $T): Int = {
+          val ${t.hash._1} = passedInObjectToHash
+          ${t.hash._2}
+        }
+
+        def get(from: _root_.java.nio.ByteBuffer): _root_.scala.util.Try[(_root_.java.nio.ByteBuffer, $T)] = {
+          val ${t.get._1} = from
+          _root_.scala.util.Try {
```
Comment (johnynek) [inline scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/TreeOrderedBuf.scala]: might as well inline the try { } catch to be a bit faster.

File: scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/TreeOrderedBuf.scala [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+        private def innerMemCompare(x: $T, y: $T): Int = {
+          val ${t.compare._1} = x
+          val ${t.compare._2} = y
+          ${t.compare._3}
+          return 0
+        }
+
+         def compare(x: $T, y: $T): Int = {
+          val tmp = innerMemCompare(x, y)
```
Comment (johnynek) [inline scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/TreeOrderedBuf.scala]: this seems wasteful, if actually all of the compares except like short and byte are doing this. Seems like we should special case those.

File: scalding-macros/src/test/scala/com/twitter/scalding/macros/MacroOrderingProperties.scala [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    implicit val localOrdering = Ordering.Option(Ordering.Option(Ordering.Int))
+    check[Option[Option[Int]]]
+  }
+
+  test("Test out TestCC") {
+    import TestCC._
+    check[TestCC](implicitly[Arbitrary[TestCC]], Ordering.by(t => TestCC.unapply(t)), implicitly)
+  }
+
+  test("Test out (Int, Int)") {
```
Comment (johnynek) [inline scalding-macros/src/test/scala/com/twitter/scalding/macros/MacroOrderingProperties.scala]: what about a class that specifically exercises Product?

```scala
class MyData(override val _1: Int, override val _2: Option[Long]) extends Product2[Int, Option[Long]]
```
just to be sure the tuple case is not picking up case class version.

File: scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/TreeOrderedBuf.scala (hunk 8 above)
Comment (ianoc) [inline scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/TreeOrderedBuf.scala]: How do you mean ? the inner function is wasteful? or trying to do the clamp?

Comment (johnynek) [inline scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/TreeOrderedBuf.scala]: I just mean most of the time, innerMemCompare will already be -1, 0, 1. Then we are going to go through the branching again. Maybe we can only do this for the explicitly needed cases (Short? Byte?)

**Comment by johnynek** [comment]
What's the story here? What is the reviewable branch?

**Comment by ianoc** [comment]
Stuck with our internal structures failing in MR jobs really from macro generated. I have a big test suite passing but I'd like to get some green MR runs before its worth doing much reviewing

**Comment by johnynek** [comment]
This is a dead-end branch right? It seems like #1165 is the latest one. Can we close this one, Ian?

**Comment by ianoc** [comment]
Yep

**Pull request closed without merging**